network/openssl_privatekey.py

```diff
@@ -0,0 +1,225 @@
+        self.path = module.params['path']
+        self.changed = True
+
+    @property
+    def file_path(self):
+        return os.path.join(self.path, self.name)
+   
+    def generate(self):
+        '''Generate a keypair.'''
+
+        if not self.force and os.path.exists(self.file_path):
+            self.changed = False
+            return
+
+        self.privatekey = crypto.PKey()
+
+        if self.type == 'RSA':
+            crypto_type = crypto.TYPE_RSA
+        else:
+            crypto_type = crypto.TYPE_DSA
+
+        try:
+            self.privatekey.generate_key(crypto_type, self.size)
+            privatekey_file = open(self.file_path, 'w')
+            privatekey_file.write(crypto.dump_privatekey(crypto.FILETYPE_PEM, self.privatekey))
+            privatekey_file.close()
+        except TypeError:
+            raise PrivateKeyError()
+        except ValueError:
+            raise PrivateKeyError()
+        except IOError:
+            try:
+                os.remove(self.file_path)
+            except OSError:
+                pass
+            raise PrivateKeyError()
+
+    def remove(self):
+        '''Remove the private key from the filesystem.'''
+        if os.path.exists(self.file_path):
+            try:
+                os.remove(self.file_path)
+            except OSError:
+                raise PrivateKeyError()
+        else:
+            self.changed = False
+
+    def dump(self):
+        '''Serialize the object into a dictionnary.'''
+
+        result = {
+            'size': self.size,
+            'type': self.type,
+            'privatekey': self.file_path,
+            'changed': self.changed,
+        }
+
+        return result
+        
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            state=dict(default='present', choices=['present', 'absent'], type='str'),
+            name=dict(required=True, type='str'),
+            size=dict(default=4096, type='int'),
+            type=dict(default='RSA', choices=['RSA', 'DSA'], type='str'),
+            force=dict(default=False, type='bool'),
+            path=dict(required=True, type='path'),
+        )
+    )
+
+    if not pyopenssl_found:
+        module.fail_json(msg="the python pyOpenSSL module is required")
+
+    path = module.params['path']
+    if not os.path.isdir(path):
+        module.fail_json(name=path, msg='The directory %s does not exist or the file is not a directory' % path)
+
+    private_key = PrivateKey(module)
+    if private_key.state == 'present':
+        try:
+            private_key.generate()
+        except PrivateKeyError:
+            module.fail_json(name=path, msg='An error occured while generating the SSL private key')
+    else:
+        try:
+            private_key.remove()
+        except PrivateKeyError:
+            module.fail_json(name=path, msg='An error occured while removing the SSL private key')
+
+    result = private_key.dump()
+
+    module.exit_json(**result)
```

Comment:
It would be nice if you also supported check_mode natively (which would basically just be checking whether a file exists or not).  Currently in --check mode the user will always see "skipped".